app/src/main/java/com/example/drawingApp/ColorSlider.kt

```diff
@@ -0,0 +1,125 @@
+package com.example.drawingApp
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.Paint
+import android.util.AttributeSet
+import android.view.MotionEvent
+import android.view.View
+import android.view.ViewGroup
+import androidx.core.view.children
+import kotlin.math.max
+import kotlin.math.min
+
+
+@SuppressLint("ClickableViewAccessibility")
+class ColorSlider @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0
+) : View(context, attrs, defStyle) {
+
+    private var held = false
+    var hue: Float = 180f
+
+    /**
+     * called whenever the hue value is changed,
+     * use to set hue value or to use the hue value that has been changed.
+     */
+    var onHueChange: ((Float) -> Unit)? = null
+
+    /**
+     * called on motionEvent Down
+     * used to prevent dragging
+     */
+    var onDown: (() -> Unit)? = null
+
+    /**
+     * called on motionEvent cancel or up
+     * used to allow dragging
+     */
+    var onUp: (() -> Unit)? = null
+
+    init {
+        setOnTouchListener { _, motionEvent ->
+            val hueCoordinate = max(min((360f - (motionEvent.y / height) * 360f), 360f), 0f)
+            when (motionEvent.action) {
+                MotionEvent.ACTION_DOWN -> {
+                    held = true
+                    onDown?.invoke()
+                    (this.parent as? ViewGroup)?.children?.forEach { it.clearFocus() }
+                    sliderSetHue(hueCoordinate)
+                    onHueChange?.invoke(hueCoordinate)
+                }
+                MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
+                    held = false
+                    onUp?.invoke()
+                }
+                MotionEvent.ACTION_MOVE -> {
+                    if (held) {
+                        sliderSetHue(hueCoordinate)
+                        onHueChange?.invoke(hueCoordinate)
+                    }
+                }
+            }
+            true
+        }
+    }
+
+    fun sliderSetHue(newHue: Float) {
+        hue = newHue
+        invalidate()
+    }
+
+    private val selectionPaint = Paint().apply {
+        strokeWidth = 5f
+        style = Paint.Style.STROKE
+        color = Color.WHITE
+    }
+
+    private fun getBitmap(): Bitmap? {
+        return if (sliderBitmap == null) {
+            sliderBitmap = generateBitmap()
+            sliderBitmap
+        } else sliderBitmap
+    }
+
+    private fun generateBitmap(): Bitmap {
+        var startHue = 360f
```

Comment:
360 should be constant

Reply:
fair enough, also incorporated a min hue constant